File: genie-backend-python/src/main/python/com/jd/genie/service/impl/multi_agent_service_impl.py
"""
Multi-agent service implementation.
"""

import asyncio
import json
import logging
import time
from typing import Any, Dict, List, Optional

import httpx
from fastapi.responses import StreamingResponse

from ..i_multi_agent_service import IMultiAgentService
from ...agent.enums.agent_type import AgentType
from ...agent.enums.auto_bots_result_status import AutoBotsResultStatus
from ...agent.enums.response_type_enum import ResponseTypeEnum
from ...config.genie_config import GenieConfig
from ...handler.agent_response_handler import AgentResponseHandler
from ...model.dto.auto_bots_result import AutoBotsResult
from ...model.multi.event_result import EventResult
from ...model.req.agent_request import AgentRequest
from ...model.req.gpt_query_req import GptQueryReq
from ...model.response.agent_response import AgentResponse
from ...model.response.gpt_process_result import GptProcessResult
from ...util.chatei_utils import ChateiUtils

logger = logging.getLogger(__name__)


class MultiAgentServiceImpl(IMultiAgentService):
    """Implementation of multi-agent service."""
    
    def __init__(
        self, 
        genie_config: GenieConfig,
        handler_map: Dict[AgentType, AgentResponseHandler]
    ):
        """
        Initialize with dependencies.
        
        Args:
            genie_config: Genie configuration
            handler_map: Map of agent types to their handlers
        """
        self.genie_config = genie_config
        self.handler_map = handler_map
    
    async def search_for_agent_request(
        self, 
        gpt_query_req: GptQueryReq, 
        streaming_response: StreamingResponse
    ) -> AutoBotsResult:
        """
        Process agent request with streaming response.
        
        Args:
            gpt_query_req: GPT query request
            streaming_response: Streaming response for SSE
            
        Returns:
            AutoBots result
        """
        agent_request = self._build_agent_request(gpt_query_req)
        logger.info(f"{gpt_query_req.request_id} start handle Agent request: {json.dumps(agent_request.__dict__)}")
        
        try:
            await self._handle_multi_agent_request(agent_request, streaming_response)
        except Exception as e:
            logger.error(
                f"{gpt_query_req.request_id}, error in requestMultiAgent, "
                f"deepThink: {gpt_query_req.deep_think}, errorMsg: {str(e)}", 
                exc_info=e
            )
            raise e
        finally:
            logger.info(
                f"{gpt_query_req.request_id}, agent.query.web.singleRequest end, "
                f"requestId: {json.dumps(gpt_query_req.__dict__)}"
            )
        
        return ChateiUtils.to_auto_bots_result(agent_request, AutoBotsResultStatus.LOADING.name)
    
    async def _handle_multi_agent_request(
        self, 
        auto_req: AgentRequest, 
        streaming_response: StreamingResponse
    ) -> None:
        """
        Handle multi-agent request with HTTP streaming.
        
        Args:
            auto_req: Agent request
            streaming_response: Streaming response for SSE
        """
        start_time = time.time() * 1000  # Convert to milliseconds
        
        # Build HTTP request
        url = "http://127.0.0.1:8080/AutoAgent"
        headers = {"Content-Type": "application/json"}
        data = json.dumps(auto_req.__dict__)
        
        logger.info(f"{auto_req.request_id} agentRequest: {data}")
        
        timeout = httpx.Timeout(
            connect=60.0,
            read=self.genie_config.sse_client_read_timeout,
            write=1800.0,
            pool=self.genie_config.sse_client_connect_timeout
        )
        
        async with httpx.AsyncClient(timeout=timeout) as client:
            try:
                async with client.stream("POST", url, headers=headers, content=data) as response:
                    await self._process_stream_response(
                        response, auto_req, streaming_response, start_time
                    )
            except Exception as e:
                logger.error(f"Request failed: {str(e)}", exc_info=e)
                raise
    
    async def _process_stream_response(
        self,
        response: httpx.Response,
        auto_req: AgentRequest,
        streaming_response: StreamingResponse,
        start_time: float
    ) -> None:
        """
        Process streaming response from the agent service.
        
        Args:
            response: HTTP response stream
            auto_req: Agent request
            streaming_response: Streaming response for SSE
            start_time: Request start time in milliseconds
        """
        agent_resp_list: List[AgentResponse] = []
        event_result = EventResult()
        
        if not response.is_success:
            error_content = await response.aread()
            logger.error(f"{auto_req.request_id}, response body is failed: {error_content.decode()}")
            return
        
        async for line_bytes in response.aiter_lines():
            line = line_bytes.decode('utf-8')
            
            if not line.startswith("data:"):
                continue
            
            data = line[5:]  # Remove "data:" prefix
            
            if data == "[DONE]":
                logger.info(f"{auto_req.request_id} data equals with [DONE]: {data}")
                break
            
            if data.startswith("heartbeat"):
                result = self._build_heartbeat_data(auto_req.request_id)
                await streaming_response.send(json.dumps(result.__dict__))
                logger.info(f"{auto_req.request_id} heartbeat-data: {data}")
                continue
            
            logger.info(f"{auto_req.request_id} recv from autocontroller: {data}")
            
            try:
                agent_response = AgentResponse(**json.loads(data))
                agent_type = AgentType.from_code(auto_req.agent_type)
                handler = self.handler_map.get(agent_type)
                
                if handler:
                    result = await handler.handle(auto_req, agent_response, agent_resp_list, event_result)
                    if result:
                        await streaming_response.send(json.dumps(result.__dict__))
                        
                        if result.finished:
                            # Record task execution time
                            total_time = time.time() * 1000 - start_time
                            logger.info(f"{auto_req.request_id} task total cost time: {total_time}ms")
                            break
                            
            except Exception as e:
                logger.error(f"Error processing stream data: {str(e)}", exc_info=e)
    
    def _build_agent_request(self, req: GptQueryReq) -> AgentRequest:
        """
        Build agent request from GPT query request.
        
        Args:
            req: GPT query request
            
        Returns:
            Agent request
        """
        request = AgentRequest()
        request.request_id = req.trace_id
        request.erp = req.user
        request.query = req.query
        request.agent_type = 5 if req.deep_think == 0 else 3
        
        if request.agent_type == 3:
            request.sop_prompt = self.genie_config.genie_sop_prompt
        else:
            request.sop_prompt = ""
            
        if request.agent_type == 5:
            request.base_prompt = self.genie_config.genie_base_prompt
        else:
            request.base_prompt = ""
            
        request.is_stream = True
        request.output_style = req.output_style
        
        return request
    
    def _build_heartbeat_data(self, request_id: str) -> GptProcessResult:
        """
        Build heartbeat data for SSE.
        
        Args:
            request_id: Request identifier
            
        Returns:
            GPT process result for heartbeat
        """
        result = GptProcessResult()
        result.finished = False
        result.status = "success"
        result.response_type = ResponseTypeEnum.TEXT.name
        result.response = ""
        result.response_all = ""
        result.use_times = 0
        result.use_tokens = 0
        result.req_id = request_id
        result.package_type = "heartbeat"
        result.encrypted = False
        
        return result

File: genie-backend/src/main/java/com/jd/genie/service/impl/MultiAgentServiceImpl.java
package com.jd.genie.service.impl;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.jd.genie.agent.enums.AgentType;
import com.jd.genie.agent.enums.AutoBotsResultStatus;
import com.jd.genie.agent.enums.ResponseTypeEnum;
import com.jd.genie.config.GenieConfig;
import com.jd.genie.handler.AgentResponseHandler;
import com.jd.genie.model.dto.AutoBotsResult;
import com.jd.genie.model.multi.EventResult;
import com.jd.genie.model.req.AgentRequest;
import com.jd.genie.model.req.GptQueryReq;
import com.jd.genie.model.response.AgentResponse;
import com.jd.genie.model.response.GptProcessResult;
import com.jd.genie.service.IMultiAgentService;
import com.jd.genie.util.ChateiUtils;
import lombok.extern.slf4j.Slf4j;
import okhttp3.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

@Slf4j
@Component
public class MultiAgentServiceImpl implements IMultiAgentService {
    @Autowired
    private GenieConfig genieConfig;
    @Autowired
    private Map<AgentType, AgentResponseHandler> handlerMap;
    @Autowired
    private OkHttpClient okHttpClient;

    @Override
    public AutoBotsResult searchForAgentRequest(GptQueryReq gptQueryReq, SseEmitter sseEmitter) {
        AgentRequest agentRequest = buildAgentRequest(gptQueryReq);
        log.info("{} start handle Agent request: {}", gptQueryReq.getRequestId(), JSON.toJSONString(agentRequest));
        try {
            handleMultiAgentRequest(agentRequest, sseEmitter);
        } catch (Exception e) {
            log.error("{}, error in requestMultiAgent, deepThink: {}, errorMsg: {}", gptQueryReq.getRequestId(), gptQueryReq.getDeepThink(), e.getMessage(), e);
            throw e;
        } finally {
            log.info("{}, agent.query.web.singleRequest end, requestId: {}", gptQueryReq.getRequestId(), JSON.toJSONString(gptQueryReq));
        }

        return ChateiUtils.toAutoBotsResult(agentRequest, AutoBotsResultStatus.loading.name());
    }

    public void handleMultiAgentRequest(AgentRequest autoReq, SseEmitter sseEmitter) {
        long startTime = System.currentTimeMillis();
        Request request = buildHttpRequest(autoReq);
        log.info("{} agentRequest:{}", autoReq.getRequestId(), JSON.toJSONString(request));
        
        // 使用单例 OkHttpClient，避免重复创建，提高性能和资源利用率
        okHttpClient.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                log.error("{} HTTP request failed", autoReq.getRequestId(), e);
                try {
                    sseEmitter.completeWithError(e);
                } catch (Exception ex) {
                    log.error("{} Failed to send error to SSE emitter", autoReq.getRequestId(), ex);
                }
            }

            @Override
            public void onResponse(Call call, Response response) {
                List<AgentResponse> agentRespList = new ArrayList<>();
                EventResult eventResult = new EventResult();
                ResponseBody responseBody = response.body();
                if (responseBody == null) {
                    log.error("{} auto agent empty response body", autoReq.getRequestId());
                    return;
                }

                // 使用 try-with-resources 确保流被正确关闭，避免资源泄露
                try (BufferedReader reader = new BufferedReader(
                        new InputStreamReader(responseBody.byteStream()))) {
                    
                    if (!response.isSuccessful()) {
                        log.error("{} HTTP response failed, status code: {}", 
                                autoReq.getRequestId(), response.code());
                        return;
                    }

                    String line;
                    while ((line = reader.readLine()) != null) {
                        if (!line.startsWith("data:")) {
                            continue;
                        }

                        String data = line.substring(5);
                        if (data.equals("[DONE]")) {
                            log.info("{} data equals with [DONE]", autoReq.getRequestId());
                            break;
                        }

                        if (data.startsWith("heartbeat")) {
                            GptProcessResult result = buildHeartbeatData(autoReq.getRequestId());
                            sseEmitter.send(result);
                            log.info("{} heartbeat-data: {}", autoReq.getRequestId(), data);
                            continue;
                        }

                        log.info("{} recv from autocontroller: {}", autoReq.getRequestId(), data);
                        AgentResponse agentResponse = JSON.parseObject(data, AgentResponse.class);
                        AgentType agentType = AgentType.fromCode(autoReq.getAgentType());
                        AgentResponseHandler handler = handlerMap.get(agentType);
                        GptProcessResult result = handler.handle(autoReq, agentResponse, agentRespList, eventResult);
                        sseEmitter.send(result);
                        if (result.isFinished()) {
                            // 记录任务执行时间
                            log.info("{} task total cost time:{}ms", autoReq.getRequestId(), System.currentTimeMillis() - startTime);
                            sseEmitter.complete();
                        }
                    }
                } catch (Exception e) {
                    log.error("{} Error processing SSE response", autoReq.getRequestId(), e);
                    try {
                        sseEmitter.completeWithError(e);
                    } catch (Exception ex) {
                        log.error("{} Failed to send error to SSE emitter", autoReq.getRequestId(), ex);
                    }
                } finally {
                    // 确保 Response 被关闭
                    response.close();
                }
            }
        });
    }

    private Request buildHttpRequest(AgentRequest autoReq) {
        String reqId = autoReq.getRequestId();
        autoReq.setRequestId(autoReq.getRequestId());
        String url = "http://127.0.0.1:8080/AutoAgent";
        RequestBody body = RequestBody.create(
                MediaType.parse("application/json"),
                JSONObject.toJSONString(autoReq)
        );
        autoReq.setRequestId(reqId);
        return new Request.Builder().url(url).post(body).build();
    }

    private GptProcessResult buildDefaultAutobotsResult(AgentRequest autoReq, String errMsg) {
        GptProcessResult result = new GptProcessResult();
        boolean isRouter = AgentType.ROUTER.getValue().equals(autoReq.getAgentType());
        if (isRouter) {
            result.setStatus("success");
            result.setFinished(true);
            result.setResponse(errMsg);
            result.setTraceId(autoReq.getRequestId());
        } else {
            result.setResultMap(new HashMap<>());
            result.setStatus("failed");
            result.setFinished(true);
            result.setErrorMsg(errMsg);
        }
        return result;
    }

    private AgentRequest buildAgentRequest(GptQueryReq req) {
        AgentRequest request = new AgentRequest();
        request.setRequestId(req.getTraceId());
        request.setErp(req.getUser());
        request.setQuery(req.getQuery());
        request.setAgentType(req.getDeepThink() == 0 ? 5: 3);
        request.setSopPrompt(request.getAgentType() == 3 ? genieConfig.getGenieSopPrompt(): "");
        request.setBasePrompt(request.getAgentType() == 5 ? genieConfig.getGenieBasePrompt() : "");
        request.setIsStream(true);
        request.setOutputStyle(req.getOutputStyle());

        return request;
    }


    private GptProcessResult buildHeartbeatData(String requestId) {
        GptProcessResult result = new GptProcessResult();
        result.setFinished(false);
        result.setStatus("success");
        result.setResponseType(ResponseTypeEnum.text.name());
        result.setResponse("");
        result.setResponseAll("");
        result.setUseTimes(0);
        result.setUseTokens(0);
        result.setReqId(requestId);
        result.setPackageType("heartbeat");
        result.setEncrypted(false);
        return result;
    }
}
